File: rjhttp/src/main/java/com/cn/lib/basic/BaseRecyclerAdapter.java
package com.cn.lib.basic;

import android.content.Context;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.LinearLayout;

import java.util.ArrayList;
import java.util.List;

import static android.view.ViewGroup.LayoutParams.MATCH_PARENT;
import static android.view.ViewGroup.LayoutParams.WRAP_CONTENT;

/**
 * Created by admin on 2016/4/1.
 */
public abstract class BaseRecyclerAdapter<T> extends RecyclerView.Adapter<BaseRecyclerViewHolder> {
    private static final int ITEM_TYPE_HEADER = 0x1000;
    private static final int ITEM_TYPE_FOOTER = 0x2000;
    protected Context context;
    protected List<T> data;
    private int layoutResId = -1;
    private int onClickPosition = -1;
    private boolean isCancelSelectedState;
    private OnItemClickListener<T> onItemClickListener;
    private LinearLayout mHeaderLayout;
    private LinearLayout mFooterLayout;

    public BaseRecyclerAdapter(Context context, int layoutResId, boolean isCancelSelectedState) {
        this.context = context;
        this.layoutResId = layoutResId;
        this.isCancelSelectedState = isCancelSelectedState;
    }

    public BaseRecyclerAdapter(Context context, int layoutResId) {
        this(context, layoutResId, false);
    }

    public void setNewList(List<T> data) {
        this.data = data == null ? new ArrayList<T>() : new ArrayList<T>(data);
        notifyDataSetChanged();
    }

    public void setExtendList(List<? extends T> data) {
        this.data = data == null ? new ArrayList<T>() : new ArrayList<T>(data);
        notifyDataSetChanged();
    }

    public void addAll(List<T> data) {
        if (this.data == null) {
            this.data = new ArrayList<T>();
        }
        this.data.addAll(data);
        notifyDataSetChanged();
    }

    public void set(T oldElem, T newElem) {
        if (oldElem == null || newElem == null) {
            return;
        }
        set(data.indexOf(oldElem), newElem);
    }

    public void set(int indexOf, T newElem) {
        if (data == null || newElem == null) return;
        if (data.size() < 1) {
            data.add(newElem);
        } else {
            data.set(indexOf, newElem);
        }
        notifyDataSetChanged();
    }

    public T getItem(int position) {
        if (data != null && data.size() > 0 && data.size() > position) {
            return data.get(position);
        } else
            return null;
    }

    public void removeItem(int position) {
        data.remove(position);
        notifyItemRemoved(position);
    }

    public void removeItem(T item) {

        removeItem(data.indexOf(item));
    }


    public List<T> getData() {
        return data == null ? this.data = new ArrayList<T>() : data;
    }

    /**
     * @return 除去头布局和脚布局长度的真实数据长度
     */
    private int getRealItemCount() {
        return data == null ? 0 : data.size();
    }


    /**
     * 插入元素操作
     */
    public void insertItem(int position, T newElem) {
        if (data == null || newElem == null) return;
        data.add(position, newElem);
        notifyItemInserted(position);
    }

    @Override
    public BaseRecyclerViewHolder onCreateViewHolder(ViewGroup parent, int viewType) {
        if (layoutResId <= 0) {
            return null;
        }
        if (viewType == ITEM_TYPE_HEADER) {
            return new HeadHolder(mHeaderLayout, context);
        }
        if (viewType == ITEM_TYPE_FOOTER) {
            return new FooterHolder(mFooterLayout, context);
        }
        View view = LayoutInflater.from(context).inflate(layoutResId, parent, false);
        return new BaseRecyclerViewHolder(view, context);
    }

    @Override
    public void onBindViewHolder(final BaseRecyclerViewHolder holder, int position) {
        if (isHeaderViewPos(position)) {
            return;
        }
        if (isFooterViewPos(position)) {
            return;
        }
        //除去头部的真实数据位置
        final int realPosition = position - getHeaderLayoutCount();
        convertView(holder, data.get(realPosition), realPosition);
        if (holder.itemView != null && onItemClickListener != null) {
            holder.itemView.setOnClickListener(new View.OnClickListener() {
                @Override
                public void onClick(View v) {
                    if (onClickPosition != realPosition) {
                        onClickPosition = realPosition;
                        notifyDataSetChanged();
                        onItemClickListener.onItemClick(onClickPosition, data.get(realPosition));
                    } else if (isCancelSelectedState) {
                        onClickPosition = -1;
                        notifyDataSetChanged();
                        onItemClickListener.onItemClick(onClickPosition, data.get(realPosition));
                    }
                }
            });
        }
    }

    private boolean isHeaderViewPos(int position) {
        return position < getHeaderLayoutCount();
    }

    private boolean isFooterViewPos(int position) {
        return position >= getHeaderLayoutCount() + getRealItemCount();
    }

    public int addHeaderView(View header) {
        return this.addHeaderView(header, -1);
    }

    public int addHeaderView(View header, int index) {
        return addHeaderView(header, index, LinearLayout.VERTICAL);
    }

    public int addHeaderView(View header, int index, int orientation) {
        if (mHeaderLayout == null) {
            mHeaderLayout = new LinearLayout(header.getContext());
            if (orientation == LinearLayout.VERTICAL) {
                mHeaderLayout.setOrientation(LinearLayout.VERTICAL);
                mHeaderLayout.setLayoutParams(new RecyclerView.LayoutParams(MATCH_PARENT, WRAP_CONTENT));
            } else {
                mHeaderLayout.setOrientation(LinearLayout.HORIZONTAL);
                mHeaderLayout.setLayoutParams(new RecyclerView.LayoutParams(WRAP_CONTENT, MATCH_PARENT));
            }
        }
        final int childCount = mHeaderLayout.getChildCount();
        if (index < 0 || index > childCount) {
            index = childCount;
        }
        mHeaderLayout.addView(header, index);
        return index;
    }

    public int addFooterView(View footer) {
        return addFooterView(footer, -1, LinearLayout.VERTICAL);
    }

    public int addFooterView(View footer, int index) {
        return addFooterView(footer, index, LinearLayout.VERTICAL);
    }

    public int addFooterView(View footer, int index, int orientation) {
        if (mFooterLayout == null) {
            mFooterLayout = new LinearLayout(footer.getContext());
            if (orientation == LinearLayout.VERTICAL) {
                mFooterLayout.setOrientation(LinearLayout.VERTICAL);
                mFooterLayout.setLayoutParams(new RecyclerView.LayoutParams(MATCH_PARENT, WRAP_CONTENT));
            } else {
                mFooterLayout.setOrientation(LinearLayout.HORIZONTAL);
                mFooterLayout.setLayoutParams(new RecyclerView.LayoutParams(WRAP_CONTENT, MATCH_PARENT));
            }
        }
        final int childCount = mFooterLayout.getChildCount();
        if (index < 0 || index > childCount) {
            index = childCount;
        }
        mFooterLayout.addView(footer, index);
        if (mFooterLayout.getChildCount() == 1) {
            int position = getFooterViewPosition();
            if (position != -1) {
                notifyItemInserted(position);
            }
        }
        return index;
    }

    /*根据位置来返回不同的item类型*/
    @Override
    public int getItemViewType(int position) {
        if (isHeaderViewPos(position)) {
            return ITEM_TYPE_HEADER;
        } else if (isFooterViewPos(position)) {
            return ITEM_TYPE_FOOTER;
        }
        return position - getHeaderLayoutCount();
    }

    @Override
    public int getItemCount() {
        return getHeaderLayoutCount() + getFootersCount() + getRealItemCount();
    }

    public int getHeaderLayoutCount() {
        if (mHeaderLayout == null || mHeaderLayout.getChildCount() == 0) {
            return 0;
        }
        return 1;
    }

    public int getFootersCount() {
        if (mFooterLayout == null || mFooterLayout.getChildCount() == 0) {
            return 0;
        }
        return 1;
    }

    /**
     * 实现逻辑代码
     *
     * @param holder
     * @param t
     * @param realPosition
     */
    public abstract void convertView(BaseRecyclerViewHolder holder, T t, int realPosition);

    public void setOnItemClickListener(OnItemClickListener<T> onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    public void setDefaultSelectPosition(int onClickPosition) {
        this.onClickPosition = onClickPosition;
    }

    public int getFooterViewPosition() {
        return getHeaderLayoutCount() + data.size();
    }

    public interface OnItemClickListener<T> {
        void onItemClick(int position, T obj);
    }

    private class HeadHolder extends BaseRecyclerViewHolder {

        private HeadHolder(View itemView, Context context) {
            super(itemView, context);
        }
    }

    private class FooterHolder extends BaseRecyclerViewHolder {

        private FooterHolder(View itemView, Context context) {
            super(itemView, context);
        }
    }
}

File: rjhttp/src/main/java/com/cn/lib/basic/BaseRecyclerAdapter.kt
package com.cn.lib.basic

import android.content.Context
import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.LinearLayout

import java.util.ArrayList

import android.view.ViewGroup.LayoutParams.MATCH_PARENT
import android.view.ViewGroup.LayoutParams.WRAP_CONTENT

/**
 * Created by admin on 2016/4/1.
 */
abstract class BaseRecyclerAdapter<T> @JvmOverloads constructor(protected var context: Context, private val layoutResId: Int, private val isCancelSelectedState: Boolean = false) : RecyclerView.Adapter<BaseRecyclerViewHolder>() {
    protected var data: MutableList<T> = mutableListOf()
    private var onClickPosition = -1
    private var onItemClickListener: OnItemClickListener<T>? = null
    private val mHeaderLayout: LinearLayout by lazy {
        LinearLayout(context)
    }
    private val mFooterLayout: LinearLayout by lazy {
        LinearLayout(context)
    }

    /**
     * @return 除去头布局和脚布局长度的真实数据长度
     */
    private fun getRealItemCount(): Int {
        return data?.size ?: 0
    }

    fun getHeaderLayoutCount(): Int {
        return if (mHeaderLayout.childCount == 0) 0 else 1
    }

    fun getFooterLayoutCount(): Int {
        return if (mFooterLayout.childCount == 0) 0 else 1
    }

    fun getFooterViewPosition(): Int {
        return getHeaderLayoutCount() + data.size
    }

    fun setNewList(data: MutableList<T>) {
        this.data = data
        notifyDataSetChanged()
    }

    fun addAll(data: List<T>) {
        this.data.addAll(data)
        notifyDataSetChanged()
    }

    fun clear() {
        this.data.clear()
        notifyDataSetChanged()
    }

    operator fun set(oldElem: T, newElem: T) {
        val index = data.indexOf(oldElem)
        set(index, newElem)
    }

    operator fun set(indexOf: Int = -1, newElem: T) {
        if (indexOf > 0) {
            this.data[indexOf] = newElem
        } else {
            this.data.add(newElem)
        }
        notifyDataSetChanged()
    }

    fun getItem(position: Int): T? {
        return data.let {
            if (it.size > position)
                it[position]
            else
                null
        }
    }

    fun removeItem(position: Int = -1) {
        if (position < data.size && position > -1) {
            this.data.removeAt(position)
            notifyItemRemoved(position)
        }
    }

    fun removeItem(item: T) {
        val index = this.data.indexOf(item)
        if (index > -1) {
            removeItem(index)
        }
    }

    fun getList(): MutableList<T> {
        return data
    }

    /**
     * 插入元素操作
     */
    fun insertItem(position: Int, newElem: T) {
        this.data.add(position, newElem)
        notifyItemInserted(position)
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): BaseRecyclerViewHolder {
        if (layoutResId <= 0) {
            throw RuntimeException("layout resource id is null")
        }
        if (viewType == ITEM_TYPE_HEADER) {
            return HeadHolder(mHeaderLayout, context)
        }
        if (viewType == ITEM_TYPE_FOOTER) {
            return FooterHolder(mFooterLayout, context)
        }
        val view = LayoutInflater.from(context).inflate(layoutResId, parent, false)
        return BaseRecyclerViewHolder(view, context)
    }

    override fun onBindViewHolder(holder: BaseRecyclerViewHolder, position: Int) {
        if (isHeaderViewPos(position)) {
            return
        }
        if (isFooterViewPos(position)) {
            return
        }
        //除去头部的真实数据位置
        val realPosition = position - getHeaderLayoutCount()
        convertView(holder, data[realPosition], realPosition)
        if (holder.itemView != null && onItemClickListener != null) {
            holder.itemView.setOnClickListener {
                if (onClickPosition != realPosition) {
                    onClickPosition = realPosition
                    notifyDataSetChanged()
                    onItemClickListener?.onItemClick(onClickPosition, data[realPosition])
                } else if (isCancelSelectedState) {
                    onClickPosition = -1
                    notifyDataSetChanged()
                    onItemClickListener?.onItemClick(onClickPosition, data[realPosition])
                }
            }
        }
    }

    private fun isHeaderViewPos(position: Int): Boolean {
        return position < getHeaderLayoutCount()
    }

    private fun isFooterViewPos(position: Int): Boolean {
        return position >= getHeaderLayoutCount() + getRealItemCount()
    }

    fun addHeaderView(header: View): Int {
        return this.addHeaderView(header, -1)
    }

    private fun addHeaderView(header: View, index: Int, orientation: Int = LinearLayout.VERTICAL): Int {
        var addIndex = index
        if (orientation == LinearLayout.VERTICAL) {
            mHeaderLayout.orientation = LinearLayout.VERTICAL
            mHeaderLayout.layoutParams = RecyclerView.LayoutParams(MATCH_PARENT, WRAP_CONTENT)
        } else {
            mHeaderLayout.orientation = LinearLayout.HORIZONTAL
            mHeaderLayout.layoutParams = RecyclerView.LayoutParams(WRAP_CONTENT, MATCH_PARENT)
        }
        val childCount = mHeaderLayout.childCount
        if (addIndex < 0 || addIndex > childCount) {
            addIndex = childCount
        }
        mHeaderLayout.addView(header, addIndex)
        return addIndex
    }

    @JvmOverloads
    fun addFooterView(footer: View, index: Int = -1, orientation: Int = LinearLayout.VERTICAL): Int {
        var addIndex = index
        if (orientation == LinearLayout.VERTICAL) {
            mFooterLayout.orientation = LinearLayout.VERTICAL
            mFooterLayout.layoutParams = RecyclerView.LayoutParams(MATCH_PARENT, WRAP_CONTENT)
        } else {
            mFooterLayout.orientation = LinearLayout.HORIZONTAL
            mFooterLayout.layoutParams = RecyclerView.LayoutParams(WRAP_CONTENT, MATCH_PARENT)
        }
        val childCount = mFooterLayout.childCount
        if (addIndex < 0 || addIndex > childCount) {
            addIndex = childCount
        }
        mFooterLayout.addView(footer, addIndex)
        if (mFooterLayout.childCount == 1) {
            val position = getFooterViewPosition()
            if (position != -1) {
                notifyItemInserted(position)
            }
        }
        return addIndex
    }

    fun clearFooterViews() {
        mFooterLayout.removeAllViews()
    }

    fun clearHeaderViews() {
        mHeaderLayout.removeAllViews()

    }

    /*根据位置来返回不同的item类型*/
    override fun getItemViewType(position: Int): Int {
        if (isHeaderViewPos(position)) {
            return ITEM_TYPE_HEADER
        } else if (isFooterViewPos(position)) {
            return ITEM_TYPE_FOOTER
        }
        return position - getHeaderLayoutCount()
    }

    override fun getItemCount(): Int {
        return getHeaderLayoutCount() + getFooterLayoutCount() + getRealItemCount()
    }

    /**
     * 实现逻辑代码
     *
     * @param holder
     * @param t
     * @param realPosition
     */
    abstract fun convertView(holder: BaseRecyclerViewHolder, t: T, realPosition: Int)

    fun setOnItemClickListener(onItemClickListener: OnItemClickListener<T>) {
        this.onItemClickListener = onItemClickListener
    }

    fun setDefaultSelectPosition(onClickPosition: Int) {
        this.onClickPosition = onClickPosition
    }

    interface OnItemClickListener<T> {
        fun onItemClick(position: Int, obj: T)
    }

    private inner class HeadHolder(itemView: LinearLayout?, context: Context) : BaseRecyclerViewHolder(itemView, context)

    private inner class FooterHolder(itemView: View?, context: Context) : BaseRecyclerViewHolder(itemView, context)

    companion object {
        private val ITEM_TYPE_HEADER = 0x1000
        private val ITEM_TYPE_FOOTER = 0x2000
    }
}
